Add support for unquoted authentication parameters

thibmeu opened the pull request:
Certain implementations of Privacy Pass do not quote authentication parameters. While there are no such example in the current authentication draft, this is a valid configuration of WWW-Authentication challenges per [RFC 9110 Section 11.2](https://www.rfc-editor.org/rfc/rfc9110#section-11.2).

This commit add supports for unquoted authentication parameters, both when parsing PrivateToken WWW-Authenticate challenge headers, and when serialising a Privave Token to such a challenge header.

thibmeu committed:
Add support for unquoted authentication parameters

Certain implementations of Privacy Pass do not quote authentication
parameters. While there are no such example in the current
authentication draft, this is a valid configuration of
WWW-Authentication challenges per [RFC 9110 Section
11.2](https://www.rfc-editor.org/rfc/rfc9110#section-11.2).

This commit add supports for unquoted authentication parameters, both
when parsing PrivateToken WWW-Authenticate challenge headers, and when
serialising a Privave Token to such a challenge header.

Lekensteyn commented on src/util.ts:
```diff
@@ -130,3 +130,48 @@ export function joinAll(a: ArrayBuffer[]): ArrayBuffer {
+    const OWS = '[ \\t]*';
+    const RWS = '[ \\t]+';
+    const BWS = OWS;
+    const qdtext = '[\\t\\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]';
```
Shouldn't this include a SP character?
```suggestion
    const qdtext = '[\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]';
```

Additionally, to reduce the number of backslashes, may I suggest:
```suggestion
    const qdtext = /[\t \x21\x23-\x5B\x5D-\x7E\x80-\xFF]/.source;
```
https://www.rfc-editor.org/rfc/rfc9110#section-5.6.4

thibmeu replied:
updated

thibmeu replied:
regarding the additional backslashes, using a regex then `.source` makes eslint complains.

Lekensteyn replied:
The space character still appears to be missing in `qdtext`

Lekensteyn commented on src/util.ts:
```diff
@@ -130,3 +130,48 @@ export function joinAll(a: ArrayBuffer[]): ArrayBuffer {
+    const RWS = '[ \\t]+';
+    const BWS = OWS;
+    const qdtext = '[\\t\\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]';
+    const quotedPair = '\\\\([ \\t' + tokenChar + '\\x21-\\x7E\\x80-\\xFF])';
```
Not sure what you were trying to define here, but the space and tabs are reversed, and `tokenChar` does not belong there:
```suggestion
    const quotedPair = /\\[\t \x21-\x7e\x80-\xff]/.source;
```
Do you also actually need the grouping `(` and `)` characters?

Based on the grammar:
```
quoted-pair = "\" ( HTAB / SP / VCHAR / obs-text )
```
From https://www.rfc-editor.org/rfc/rfc5234.html
```
         VCHAR          =  %x21-7E
                                ; visible (printing) characters
```

thibmeu replied:
updated

Lekensteyn commented on src/util.ts:
```diff
@@ -130,3 +130,48 @@ export function joinAll(a: ArrayBuffer[]): ArrayBuffer {
+    const authParam = `${token}${BWS}=${BWS}(?:${token}|${quotedString})`;
+    const authScheme = token;
+    const token68 = `[\\/+\\-\\._~${DIGIT}${ALPHA}]+=*`;
+    const challenge = `${authScheme}(?:${RWS}(?:${token68}|${authParam}(?:${OWS},${OWS}${token68}|${authParam})*))?`;
```
Spec says
```
challenge   = auth-scheme [ 1*SP ( token68 / #auth-param ) ]
```
which basically [expands](https://www.rfc-editor.org/rfc/rfc9110#abnf.extension) to:
```
challenge   = auth-scheme [ 1*SP ( token68 / (auth-param *( OWS "," OWS auth-param ) ) ) ]
```
so the `${RWS}` should be space to not match tabs, but one of more spaces. That is also assumed by the `PrivateToken ` check above. Given that more than one space can be matched, the code above should probably be updated.
```suggestion
    const challenge = `${authScheme}(?: +(?:${token68}|${authParam}(?:${OWS},${OWS}${token68}|${authParam})*))?`;
```

thibmeu replied:
updated

Lekensteyn commented on src/util.ts:
```diff
@@ -130,3 +130,48 @@ export function joinAll(a: ArrayBuffer[]): ArrayBuffer {
+            return [];
+        }
+        if (groups.challenge_optional) {
+            challengesRaw.push(groups.challenge_optional);
```
This seems rather expensive, matching the whole thing over and over again, throwing away the result.

Here is a sketch of a simplified alternative that does a single pass using a [sticky](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/sticky) RegExp flag to ensure that no substring is ignored.
```javascript
let testCases = [
    ["A, B, C", ["A", "B", "C"]],
    ["A,  B, C", ["A", "B", "C"]],
    ["A,B", ["A", "B"]],
    ["A", ["A"]],
    ["A,", []],
    ["A,Bx", []],
    [",A", []],
];

testCases.forEach(([str, expected]) => {
    let regex = /(?<skip> *, *)?(?<data>[A-Z]+)/y;
    let first = true;
    var m;
    let everythingConsumed = false;
    let matches = [];
    while ((m = regex.exec(str))) {
        if (first) {
            if (m.groups.skip) {
                break;
            }
            first = false;
        }
        matches.push(m.groups.data);
        everythingConsumed = str.length === regex.lastIndex;
    }
    if (!everythingConsumed) {
        matches = [];
    }
    //console.log(matches);
    if (matches.toString() !== expected.toString()) {
        console.log("expected", expected, "got", matches);
    }
});
```

thibmeu replied:
updated
I've slightly modified the naming and the while to scope `m` variable to the for loop.

armfazh commented on src/httpAuthScheme.ts:
```diff
@@ -180,29 +180,42 @@ export class PrivateToken {
     }
 
     static async parseMultiple(header: string): Promise<PrivateToken[]> {
-        // Consumes data:
```
if it's still accurate, let's keep the documentation

thibmeu replied:
moved back

armfazh commented on src/util.ts:
```diff
@@ -130,3 +130,52 @@ export function joinAll(a: ArrayBuffer[]): ArrayBuffer {
+        if (data) {
+            matches.push(data);
+        }
+        everythingConsumed = header.length === challengesRegex.lastIndex;
+    }
+    if (!everythingConsumed) {
```
```suggestion
    }

    // check if everything was consumed
    if (header.length !== challengesRegex.lastIndex) {
```

thibmeu replied:
I'm not sure this would work given challengeRegex is modified after the check.

armfazh commented on test/testdata/authscheme_header_v11.json:
```diff
@@ -16,5 +16,12 @@
         "max-age-1": 10,
         "token-challenge-1": "0001000e6973737565722e6578616d706c65208a3e83a33d98005d2f30bef419fa6bf4cd5c6005e36b1285bbb4ccd40fa4b383000e6f726967696e2e6578616d706c65",
         "WWW-Authenticate": "PrivateToken challenge=\"AAIADmlzc3Vlci5leGFtcGxlIIo-g6M9mABdLzC-9Bn6a_TNXGAF42sShbu0zNQPpLODAA5vcmlnaW4uZXhhbXBsZQ==\", token-key=\"MIIBUjA9BgkqhkiG9w0BAQowMKANMAsGCWCGSAFlAwQCAqEaMBgGCSqGSIb3DQEBCDALBglghkgBZQMEAgKiAwIBMAOCAQ8AMIIBCgKCAQEAyxrta2qV9bHOATpM_KsluUsuZKIwNOQlCn6rQ8DfOowSmTrxKxEZCNS0cb7DHUtsmtnN2pBhKi7pA1I-beWiJNawLwnlw3TQz-Adj1KcUAp4ovZ5CPpoK1orQwyB6vGvc ...
+    },
```
does this test vector came from the spec? if not, let's bring this one to the reference implementation

thibmeu replied:
moved.

armfazh commented on src/util.ts:
```diff
@@ -130,3 +130,52 @@ export function joinAll(a: ArrayBuffer[]): ArrayBuffer {
+//   PrivateToken challenge="def...", token-key="234...",
+//   PrivateToken challenge=ghi..., token-key=345...
+// Parse WWW-Authenticate according to RFC9110 Section 11.6.1 https://www.rfc-editor.org/rfc/rfc9110#section-11.6.1
+export function parseWWWAuthenticate(header: string): string[] {
```
it would be good to move all code related to rfc9110 to a file called rfc9110.ts

thibmeu replied:
moved

armfazh commented on test/testdata/authscheme_header_v11.json:
```diff
@@ -16,5 +16,12 @@
         "token-challenge-1": "0001000e6973737565722e6578616d706c65208a3e83a33d98005d2f30bef419fa6bf4cd5c6005e36b1285bbb4ccd40fa4b383000e6f726967696e2e6578616d706c65",
         "WWW-Authenticate": "PrivateToken challenge=\"AAIADmlzc3Vlci5leGFtcGxlIIo-g6M9mABdLzC-9Bn6a_TNXGAF42sShbu0zNQPpLODAA5vcmlnaW4uZXhhbXBsZQ==\", token-key=\"MIIBUjA9BgkqhkiG9w0BAQowMKANMAsGCWCGSAFlAwQCAqEaMBgGCSqGSIb3DQEBCDALBglghkgBZQMEAgKiAwIBMAOCAQ8AMIIBCgKCAQEAyxrta2qV9bHOATpM_KsluUsuZKIwNOQlCn6rQ8DfOowSmTrxKxEZCNS0cb7DHUtsmtnN2pBhKi7pA1I-beWiJNawLwnlw3TQz-Adj1KcUAp4ovZ5CPpoK1orQwyB6vGvc ...
+    },
+    {
```
this file shouldn't be modified

thibmeu committed:
Fix lint unsafe regexp

Correctly parse and validate RFC 9110 header, with a more complex regex
construction.

thibmeu commented:
update the commit to address @armfazh comments:
* documentation is not modified
* test vector from RFC is not modified
